### DISCUSSION TITLE
Using non-singleton service in transform

### DISCUSSION THREAD

**Discussion by leoshusar** [dotnet/yarp#1817] [Q&A] [answered]
Hi, I need to proxy login request and transform the response:
- take the session header
- generate random token
- save the session to the database with the new token
- remove the session header from response
- modify the response body, append the new token to the JSON response sent from the server

I have been successful with the `ITransformProvider` with just `IMemoryCache`, everything worked perfectly.
But when I wanted to inject my `DbContext`, I realized that transform providers are registered as singleton and thus injecting `DbContext` won't work here.

What is my best option here?
- create middleware - but the docs discourage from using middleware for what I am trying to do and say transforms are better fit for it
- inject `IServiceProvider` into `ITransformProvider` and get the `DbContext` manually in each transform func call
- any better idea?

**Comment by karelz** [comment, marked as answer]
Triage: This looks like general DI scoping problem -- can happen even if you have middleware and you need `DbContext` there.
We think that getting it manually is probably best. It should be documented somewhere in ASP.NET docs.
@halter73 @davidfowl any thoughts here?

**Reply by leoshusar** [reply]
Thank you!

Just an idea, wouldn't it be possible to add functionality that would inject dependencies like new minimal APIs do?

Something like:
```cs
context.AddResponseTransform(async (ResponseTransformContext transformContext, MyDbContext dbContext) =>
{
});
```

**Reply by halter73** [reply]
@leoshusar Make sure you use `transformContext.HttpContext.RequestServices.GetRequiredService<MyDbContext>()` instead of an `IServiceProvider` injected into your `ITransformProvider`. Using an `IServiceProvider` injected into a singleton service to get another service has the exact same problems as directly injecting the other service. You could manually create a scope, but using `RequestServices` is easier.

The minimal API like suggestion in a cool one, but at least right now, it's not as easy to inject services into `Delegate`s as it is to do with constructors and DI. I do like the idea though. We should make it easier to create APIs like that.

**Reply by leoshusar** [reply]
@halter73 thank you for the DI tip!

Yeah, I think since I'm probably the first to request this, I think minimal API like injection can wait.
Maybe in the future there will be some simple universal API for enabling injections into `Delegate`s for everyone to use.

**Comment by samsp-msft** [comment]
Possible, yes, but there is a lot behind the minimal APIs required to make it work. I'm not sure if that is worth it for this scenario.
